SDL_net for SDL 1.2 on the PSP.

superfury opened the pull request:
These are the changes I made to allow SDL_net to compile on the Minimalist PSPSDK 0.10.0

It's my first attempt at actually porting any library, but it looks fine as far as I can see (just a few header changes and missing Makefile due to autoconf not working on the devkit). It's working pretty much exactly like the other PSP ports of the libraries (SDL_gfx etc.).

Although I didn't test the actual functionality on a real PSP or emulated system (other than a quick test if it doesn't crash).
And of course platform-specific stuff like connecting to the WiFi network on the PSP is to be done by the app itself (as it has nothing to do with SDL_net directly, like on other operating systems).

sezero commented on SDLnetsys.h:
```diff
@@ -49,18 +49,24 @@
 #include <fcntl.h>
 #include <netinet/in.h>
-#ifndef __BEOS__
+#if !defined(__BEOS__) || defined(__PSP__)
```
This does not seem to be needed: Already was always including it for non-beos.

superfury replied:
You're right.

sezero commented on SDLnetsys.h:
```diff
@@ -49,18 +49,24 @@
+#include <sys/select.h> //Required for the FD_SET etc.
+#endif
+
+#if defined(__OS2__)
```
No need changing `#ifdef` to `#if defined`

superfury commented:
After those changes @sezero suggested (applied at the two later commits), the library still properly compiles.

sezero commented:
Sam: Is this OK?

slouken commented:
Yes, though it should go in the main branch of SDL_net.

superfury commented:
Also, for information, the Makefile is based on the following Makefile for the PSP from SDL2:
https://github.com/mikesart/SDL2/blob/master/Makefile.psp

Just adjusted with the required things that SDL_net needs (it's required libraries and the SDL include path). I've also added a install target to install it where SDL 1.2 is also installed.
Although sdl-config --cflags returns a rewrite of the main function, which is luckily unused by the library (it does need the include directory of it for headers though).

Also Sam, I don't know if this would work on the main branch, as it's SDL2? It's made to work properly with SDL 1.2.9 that's in the PSP devkit. So that's why I've posted it here on SDL 1.2 (for compatibility with that one, which the PSP devkit is shipped with).
Probably it would need to be added to both branches? One for the 1.2 (the (squashed) commits I've made) for the PSP devkit to be able to use, and another one on the main branch (assuming it's compatible).

madebr commented:
At some later time, is it okay to remove the `Makefile.psp` and "embed" it into the cmake script?

superfury commented:
> At some later time, is it okay to remove the `Makefile.psp` and "embed" it into the cmake script?

That's an issue with some of the older devkits, like the one I'm using for the PSP (PSPSDK 0.10.0). It doesn't offer the newer autoconfiguration (configure command etc.) nor cmake. So it requires that Makefile.psp file to get SDL_net (and SDL2_net for that matter) to be compiled and install on it.

Also, as can be seen in https://github.com/libsdl-org/SDL_net/pull/65 , I've created a version for the main branch (which has some differences because of different SDL versions and base code of the repository itself).

The one from the commits in this pull request otoh are meant for the SDL_net 1.2 branch instead (and require that branch to be compatible with the SDL 1.2-only PSP devkits).

superfury commented:
Also, a little note with the devkit I'm using:

The SDL_net headers require the definition of SDL_reinterpret_cast to function properly on the PSP. If it's not defined, as is the default case on SDL 1.2.9 my devkit ships with, the functions defined with it in the SDL_net headers will fail to compile properly when used by software (it won't affect the compilation of the SDL_net library itself, just the application using said library).
SDL 1.2.15 (where I took my application's defintions of the SDL_reinterpret_cast from, from SDL_stdinc.h) does support it, so that somewhere between those two versions of SDL 1.2.x the SDL_net library will start compiling properly.

It can be fixed, of course, by performing SDL_VERSION_ATLEAST with the values of 1.2.14 and higher. Version 1.2.14 have the required reinterpret_cast (from a quick look of the SDL_stdinc.h), while all lower versions don't have it.
So the additional code in SDL_net.h would need to be:

```
#if !SDL_VERSION_ATLEAST(1,2,14)
#ifdef __cplusplus
#define SDL_reinterpret_cast(type, expression) reinterpret_cast<type>(expression)
#else
#define SDL_reinterpret_cast(type, expression) ((type)(expression))
#endif
#endif
```

superfury commented:
Just added a commit to add the above SDL_reinterpret_cast support for SDL versions too old to support it.
Now with that added, SDL_net 1.2 can be used on the PSP without special modifications made to any source code built for it.

Is this fine and enough to be able to pull it into SDL-1.2 (with those latest commits for the SDL_reinterpret_cast added)? The matter of the main (which is SDL2) is the other pull request of course, which shouldn't need that modification (since it's requiring SDL2, it's by default new enough to support it).

You thoughts, Sam?

sezero commented:
https://github.com/libsdl-org/SDL_net/commit/c87fda1d9847c7a47b9c763e251c6411907393b6 removed all reinterpret_cast thingies so I don't know.. (reinterpret_cast was added for https://github.com/libsdl-org/SDL_net/issues/9, with reference issue https://github.com/libsdl-org/SDL/issues/85)

superfury commented:
> [c87fda1](https://github.com/libsdl-org/SDL_net/commit/c87fda1d9847c7a47b9c763e251c6411907393b6) removed all reinterpret_cast thingies so I don't know.. (reinterpret_cast was added for #9, with reference issue [libsdl-org/SDL#85](https://github.com/libsdl-org/SDL/issues/85))

It doesn't matter, in that case, the code will still run on both versions if the version check for SDL_VERSION_ATLEAST is removed. Since it's always an #define, it will be added as needed.
I've modified the latest commit to this to support that.

sezero commented:
I wonder why the reinterpret_cast thing was actually removed in SDL2 version, though: by accident or intentional?

sezero commented:
> [c87fda1](https://github.com/libsdl-org/SDL_net/commit/c87fda1d9847c7a47b9c763e251c6411907393b6) removed all reinterpret_cast thingies so I don't know.. (reinterpret_cast was added for #9, with reference issue [libsdl-org/SDL#85](https://github.com/libsdl-org/SDL/issues/85))

> I wonder why the reinterpret_cast thing was actually removed in SDL2 version, though: by accident or intentional?

@slouken, @icculus: Do you remember anything about this? If the removal of reinterpret_cast in 2.0 was intentional (why?), we can then remove the thing from 1.2 branch too.

madebr commented:
I can't speak for the reasons back in the days, but I know the use of `reinterpret_cast` is considered bad practice in c++ land.
It bypasses a lot of type safety.

superfury commented:
OK.

I've tested the SDL_net with my own application now. (Although I don't test the UDP connection, since I have no PSP software that uses it) I did test the TCP functionality with my latest code(that's application-specific) to connect to the WiFi. Then I used both the SDL_net library to send and receive bytes to and from another TCP client that connects and is connected to on a TCP port. Both establishing and receiving work without issues.

So the SDL_net (with all the changes I added for compatibility) library works without issues on the PSP. I don't see anything weird going on in the connections and they seem to be stable.

So it's ready to be pulled into the SDL_net 1.2 repository (the same is probably true for SDL2_net, but I don't have any software that uses SDL2 and SDL2_net on the PSP).

That reinterpret_cast issue is an issue not directly linked with this pull request (other than some defines for it to work properly are needed). It was already in the SDL_net 1.2 repository, so probably leave it like that (unless someone wants to rip it out later, together with all those reinterpret_cast places in the SDL_net.h header).

Sam?
